doc/logging/syslog_forward_hld.md

```diff
@@ -0,0 +1,255 @@
+
+### 1.1.1 Functional Requirements
+
+Provide configuration options for  remote syslog server, port, source-ip and VRF.
+
+### 1.1.2 Configuration and Management Requirements 
+Provide REST-API and KLISH cli commands support to configure remote-server, remote-port, source-ip and VRF using management framework.
+
+### 1.1.3 Scalability Requirements
+MAX 8 syslog servers will be supported
+### 1.1.4 Warm Boot Requirements
+
+N/A
+
+## 1.2 Design Overview
+### 1.2.1 Basic Approach
+Implement configuration of remote syslog server and add support for source-ip and VRF
+
+### 1.2.2 Container
+Changes are part of management framework container, redis-db and hostcfgd
+
+### 1.2.3 SAI Overview
+N/A
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+User can use this feature to forward syslog messages to remote server
+## 2.2 Functional Description
+User can forward syslog messages generated on the switch to centralized remote syslog server for better log analysis by tools and to save space on the switch. 
+
+# 3 Design
+## 3.1 Overview
+Whenever user changes syslog configuration, CONFIG_DB notifies hostcfgd to start syslog configuration script which in turn writes syslog server configuration to /etc/rsyslog.conf file using jinja2 template and restart the syslog service.
+## 3.2 DB Changes
+Describe changes to existing DBs or any new DB being added.
+### 3.2.1 CONFIG DB
+
+**3.2.1.1 SYSLOG_SERVER Table Schema**
+
+```
+;Syslog server configuration in the system
+;Key
+ipaddress          = IPAddress; (IPv4 or IPv6)
+;Attributes
+remote-port        = 1*5DIGIT
+source-address     = IPAddress; (IPv4 or IPv6)
+vrf-name           = vrfName
+```
+
+
+
+### 3.2.2 APP DB
+
+N/A
+
+### 3.2.3 STATE DB
+
+N/A
+
+### 3.2.4 ASIC DB
+
+N/A
+
+### 3.2.5 COUNTER DB
+
+N/A
+
+## 3.3 Switch State Service Design
+### 3.3.1 Orchestration Agent
+### 3.3.2 Other Process 
+N/A
+
+## 3.4 SyncD
+N/A
+
+## 3.5 SAI
+N/A
+
+## 3.6 User Interface
+### 3.6.1 Data Models
+Following yang models are used for syslog server configuration.
+
+openconfig-system-logging.yang
+
+openconfig-system-ext.yang
+
+sonic-system-logging.yang
+
+```diff
+|  +--rw remote-servers
+     |     +--rw remote-server* [host]
+     |        +--rw host      -> ../config/host
+     |        +--rw config
+     |        |  +--rw host?                  oc-inet:host
+     |        |  +--rw source-address?        oc-inet:ip-address
+     |        |  +--rw remote-port?           oc-inet:port-number
+     |        |  +--rw oc-sys-ext:vrf-name?   string
+     |        +--ro state
+     |           +--ro host?             oc-inet:host
+     |           +--ro source-address?   oc-inet:ip-address
+     |           +--ro remote-port?      oc-inet:port-number
+                 +--ro oc-sys-ext:vrf-name?   string
+```
+
+
+
+### 3.6.2 CLI
+#### 3.6.2.1 Configuration Commands
+
+```
+sonic(config)# logging server host 10.59.142.126
+sonic(config)# logging server host 10.59.143.28 source-ip 10.10.10.1 vrf-name Vrf1
```

Comment:
Can we add remote-port configuration in KLISH?

Reply:
Looks like it's already added but with wrong description, please correct it.

sonic(config)# logging server host 1::1
  remote-port  RADIUS server's authentication port
  source-ip    Source ip address
  vrf-name     VRF to use

Reply:
Please add the CLI command syntax and examples.

Reply:
Can we use source-interface instead of source-ip to match behavior of various NOS vendors today? we're planning to migrate to source-interface instead source-ip for other IP services as well.

Reply:
sonic(config)# logging server host 1::1 vrf-name
  Name of VRF (Max size 32, (mgmt, or prefixed by Vrf_)
>> Vrf name cant be 32, please follow other vrf-name syntax and fix accordingly.

Reply:
Remote host should accept "Hostname" as well for configuration, please make the change.

Reply:
for 2 and 3) we will take care.
for 4)I think source-ip is better. Even if we use source interface, we have to get its ip address and  use it as source-ip while sending out packtes. What if interface has more than one ipadddress?. What logic should be followed to choose ip?. this makes more complex. For simplicity source-ip is better. I think ask from customer also source-ip.
5) We will take care.
6)Not sure radius/tacacs uses hostname. If other clients uses we will also follow.

Reply:
4) source interface is better from user perspective that's what IS-CLI follows, yes, internally we have to take the IP address from the source interface and use it for IP services. why do you expect more than one IPv4/IPv6 address on an interface that we want to use for any IP services, we can discuss in the email.

Reply:
I'll ask Jose (PLM) to comment on 4) further.